src/components/CodeView/index.spec.tsx

```diff
@@ -471,6 +495,82 @@ describe(__filename, () => {
     expect(getLinkText(false)).toEqual('View the original file.');
   });
 
+  describe('add comment button', () => {
+    it('renders an add comment button on each line', () => {
+      const AddComment = () => <button type="button">Add</button>;
+
+      const { shell, renderContent } = simulateCommentableLine({
+        addCommentButton: <AddComment />,
+        content: 'first line \nsecond line',
+        enableCommenting: true,
+      });
+
+      const line1 = renderContent(shell.at(0));
+      expect(line1.find(AddComment)).toHaveLength(1);
+
+      const line2 = renderContent(shell.at(1));
+      expect(line2.find(AddComment)).toHaveLength(1);
+    });
+
+    it('does not render add comment buttons when the feature is disabled', () => {
+      const AddComment = () => <button type="button">Add</button>;
+
+      const { shell, renderContent } = simulateCommentableLine({
+        addCommentButton: <AddComment />,
+        content: 'single line of code',
+        enableCommenting: false,
+      });
+
+      expect(renderContent(shell.at(0)).find(AddComment)).toHaveLength(0);
+    });
+  });
+
+  describe('comment list', () => {
+    it('renders a comment list', () => {
+      const version = createInternalVersion(fakeVersion);
+      const { shell } = simulateInlineCommentList({
+        content: 'single line of code',
+        enableCommenting: true,
+        version,
+      });
+
+      expect(shell).toHaveProp('addonId', version.addon.id);
+      expect(shell).toHaveProp('fileName', version.selectedPath);
+      expect(shell).toHaveProp('line', 1);
+      expect(shell).toHaveProp('versionId', version.id);
+    });
+
+    it('does not render a comment list when the feature is disabled', () => {
+      const { shell } = simulateInlineCommentList({
+        enableCommenting: false,
+      });
+
+      expect(shell).toHaveLength(0);
+    });
+
+    it('renders comment list content for each line', () => {
+      const CommentListResult = () => <div />;
+
+      const { shell, renderContent } = simulateInlineCommentList({
+        commentList: <CommentListResult />,
```

Comment:
So `simulateInlineCommentList` automatically puts `commentList` onto each line of content?

This is actually making me a bit worried about this approach with the simulators. Maybe I'm just not getting the abstraction, but it seems like we're setting up data in a way that is totally different from how the app actually works, and then testing the component to see if it looks the way we expect it to. This feels like we could be missing something in terms of how everything works together, but I'm guessing you feel that the tests that already exist for `CommentList` are good enough to cover this?

Reply:
Disclaimer: render props are really annoying for shallow rendering. They are an exception to how JSX flows and each one requires a special test helper.

> ... it seems like we're setting up data in a way that is totally different from how the app actually works

My intention was to use static types to enforce this. However, I took a closer look and was able to improve / simplify the type safety so that any edit to a `children` prop will nag the developer to fix the test code.

> This feels like we could be missing something in terms of how everything works together, but I'm guessing you feel that the tests that already exist for `CommentList` are good enough to cover this?

Yeah, my philosophy here is that we only need to test that `CommentList` gets *called* correctly. If it gets called correctly then we can assume things will work because it already has its own tests. 

Typically we'd do this by 1) finding a sub-component and asserting its props are correct and 2) relying on static analysis to catch mismatched props. 

In the case of render props, we have to go one extra step and *simulate* how the render prop will be called, based on its signature. That last part about the signature is important. We use static types to enforce it.